### PR TITLE
batches: encrypt site credentials

### PR DESCRIPTION
**Pull request by LawnGnome**
Part of #19570.

If you saw #20287, this is going to look pretty familiar: new OOB migrator, `SiteCredential` gains new getter and setter methods, that kind of thing.

To address a couple of questions I expect to come up:

#### Why add a new OOB migrator instead of modifying the user credential migrator, since we haven't released anything?

It makes reasoning about the migration and progress tracking hard: instead of querying one table, we need to query two tables and combine the results. Adding another migrator keeps this clean.

#### Why does `SiteCredential` have more new methods than `UserCredential` did?

Ultimately, this is because of how we structure our types in `enterprise/internal/batches`: the store methods and types are in different subpackages, which means that — since Go doesn't have a concept of `friend` like C++ (which is _good_) — the store methods need a bit more access into the internals of `SiteCredential` than would otherwise be ideal.

The options were either to break encapsulation (which I did anyway by making `Key` public, but would prefer not to for _actually_ sensitive data that could be mishandled), or merge the types and store, which is an obvious non-starter, since we actually went to some effort to keep them separate.

#### Does `Key` really belong on `types.SiteCredential`?

Now _that_ is a good question. It's almost like I wrote both the questions and the answers.

In #20287, I briefly flirted with the idea of adding an new `auth.Authenticator` implementation that was, essentially, an opaque authenticator that delegates back to another authenticator, but only decrypts when it absolutely has to. That felt unnecessary, so I didn't.

We've hit a similar problem here, though. The best practice guidance we have is not to decrypt on `SELECT`, but instead decrypt only when we actually need access to the data. However, that means we either have a stateful field in the `SiteCredential` type, or we go back to the original version of #20287 where we were drilling `encryption.Decrypter` arguments through 2395* layers of execution.

I think this is still the right approach, but I'm a little more shook than I was then. Feedback welcome.

<sup>\* Really not that much of an exaggeration.</sup>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,7 +20,7 @@ All notable changes to Sourcegraph are documented in this file.
 
 ### Changed
 
-- User credentials used in Batch Changes are now encrypted if encryption is enabled in the database with the `encryption.keys` config. [#19570](https://github.com/sourcegraph/sourcegraph/issues/19570)
+- User and site credentials used in Batch Changes are now encrypted if encryption is enabled in the database with the `encryption.keys` config. [#19570](https://github.com/sourcegraph/sourcegraph/issues/19570)
```
Comment (mrnugget) [inline CHANGELOG.md:23]:
```suggestion
- User and site credentials used in Batch Changes are now encrypted if encryption in the database is enabled with the `encryption.keys` config. [#19570](https://github.com/sourcegraph/sourcegraph/issues/19570)
```

**Review by mrnugget — APPROVED** [review]
I think the structure with the `Key` being set is fine. I'm not super familiar with that part of the codebase so can't speak to the ergonomics of the resulting code, but it doesn't look "obviously wrong" to me :)

File: CHANGELOG.md (hunk 1 above)
Comment (eseliger) [reply CHANGELOG.md:23]: To me Adams version sounded more correct 🤔 :D

**Review by eseliger — APPROVED** [review]
That looks alright to me. It's a bit sad we have to use `database` stuff in the `types` package, but if it makes things reasonably easier, I don't think we should change it. (And I wouldn't have a good idea how to change it either 🤷‍♂️)
Great work 🌟

File: CHANGELOG.md (hunk 1 above)
Comment (mrnugget) [reply CHANGELOG.md:23]: Adam's version read to me as if encryption is something that is enabled in the database (by updating a row) vs. encryption is used in the database.

V3:

```suggestion
- User and site credentials used in Batch Changes are now encrypted in the database if encryption is enabled with the `encryption.keys` config. [#19570](https://github.com/sourcegraph/sourcegraph/issues/19570)
```

Comment (LawnGnome) [reply CHANGELOG.md:23]: Thanks, I like that final version!

**Pull request merged**